### PR TITLE
[feat]: add endpoint for users to join the waitlist

### PR DESCRIPTION
**Pull request by Laban254**
## Description

This PR introduces a new API endpoint `/api/v1/waitlist/join` to facilitate user to join the waitlist form on the website. 
functions:

-   Validates and sanitizes incoming user data, including email format and required fields (email, full name).
-   Stores user information securely in the database upon successful validation.
-   Sends a confirmation email to the user upon successful registration.
-   Handles errors gracefully, returning appropriate status codes and messages.

## Related Issue (Link to issue ticket)

Link to the issue :  [#328](rhttps://github.com/hngprojects/hng_boilerplate_python_fastapi_web/issues/328)

## Motivation and Context

The introduction of this API endpoint addresses the need for a streamlined process for users to join the waitlist for our product/service. It enhances user experience by automating the waitlist signup process and ensuring that user data is handled securely and efficiently.

## How Has This Been Tested?

I have  tested this feature using  Postman. The tests cover various scenarios, including valid and invalid inputs,  and error handling. All tests have passed successfully.  To run the test  
` python -m pytest  tests/v1/waitlist_test.py ` 

## Screenshots:

signup
Screenshot from 2024-07-24 16-39-05
invalid email
Test
Screenshot from 2024-07-24 16-46-49
Email

Screenshot from 2024-07-24 19-20-11

## Types of changes

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-  [x]  My code follows the code style of this project.
-   [ ] My change requires a change to the documentation.
-   [ ] I have updated the documentation accordingly.
-   [x] I have read the  **CONTRIBUTING**  document.
-   [x] I have added tests to cover my changes.
-   [x] All  tests passed.

### CODE REVIEW

File: api/v1/routes/waitlist.py
Comment (joboy-dev) [inline api/v1/routes/waitlist.py:1]: Handle database operations in services

**Review by joboy-dev — CHANGES_REQUESTED** [review]
Check the comment i made
And check your test file

**Comment by Laban254** [comment]
Hello @joboy-dev  I have fixed the requested changes.

**Comment by CodewithSegNet** [comment]
Looks good. Please wait for a mentor to review it.

File: api/v1/routes/waitlist.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+from fastapi import APIRouter, HTTPException, Depends, Request
+from sqlalchemy.orm import Session
+from pydantic import BaseModel
+from api.v1.schemas.waitlist import WaitlistAddUserSchema
+from api.v1.services.waitlist_email import send_confirmation_email, add_user_to_waitlist, find_existing_user
+from api.utils.logger import logger
+from fastapi.responses import JSONResponse
+from api.db.database import get_db
+
+waitlist = APIRouter(prefix="/waitlist", tags=["Waitlist"])
```
Comment (zxenonx) [inline api/v1/routes/waitlist.py:10]:
```suggestion
waitlist = APIRouter(prefix="/waitlists", tags=["Waitlist"])
```

File: api/v1/routes/waitlist.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+class WaitlistResponse(BaseModel):
+    message: str
+
+@waitlist.post("/join", response_model=WaitlistResponse, status_code=201)
```
Comment (zxenonx) [inline api/v1/routes/waitlist.py:15]:
```suggestion
@waitlist.post("/", response_model=WaitlistResponse, status_code=201)
```

**Review by zxenonx — CHANGES_REQUESTED** [review]
Great PR @Laban254 
I left a comment

**Comment by buka4rill** [comment]
I have closed this PR, because it did not follow the convention

**Comment by Laban254** [comment]
> Great PR @Laban254 I left a comment

Thanks @zxenonx working on it.

**Pull request closed without merging**